accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,245 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+}
+```
+
+At a join point in the program, it is an error if an **untyped variable** has
+been given an **initialization inferred type** on one of the reachable paths to
+the join point and not on another reachable path, unless the variable is
+declared as `late`.  Note that it is valid for a variable to have no
+**initialization inferred type** on a path into a join point which is known to
+be unreachable.
+
+```dart
+void test(bool b) {
+  var x; // x is an untyped variable
+  if (b) {
+    x = 3;
+  } // Error, x has not been given an inferred type on all paths
+
+  late var y; // y is a late untyped variable
+  if (b) {
+    y = 3;
+  } // Ok, y has inferred type `int`
+
+  var z;
+  if (b) {
+     z = 3;
+  } else {
+    return;
+  }  // No error: z has been given a type on all reachable paths into the join
+}
+```
+
+At a join point in the program, if an **untyped variable** has been given an
+**initialization inferred type** on more than one of the reachable paths to the
+join point, the variable is treated after the join point as having the
+**initialization inferred type** equal to the upper bound of the
+**initialization inferred types** from the reachable paths into the join point.
+The **initialization inferred type** so computed is treated as a **type of
+interest** for the variable in question (including its non-nullable version if
+applicable).  The promoted type chains for the variable on the respective paths
+are intersected as usual.
+
+```dart
+void test(bool b) {
+  var x; // x is an untyped variable
+  if (b) {
+    x = 3;
+  } else {
+    x = 3.0;
+  } // x is has type num = UP(int, double)
+
+  late var y; // y is a late untyped variable
+  if (b) {
+    if (b) {
+      y = 3;
+    } else {
+      y = 3.0;
+    } // y has type num
+    y = 3.0; // y is promoted to double, since double is a type of interest
+    y = (3 as num); y is demoted to num.
+  } // y still has type num
+  y = "hello";  // Error, not an initializating write.
+
+ var z; // z is an untyped variable
+ if (b) {
+   x = 3; // x is promoted to int
+ } else {
+   x = 3.0;
+   return;
+ }
+ x.isEven; // No error, only one reachable path
+}
+```
+
+It is an error to read an **untyped variable** when it has no **initialization
+inferred type**.
+
+
+```dart
+void test(bool b) {
+  var x; // x is an untyped variable
+  print(x); // Error, no initialization inferred type
+
+  late var y; // y is a late untyped variable
+  var f = () => print(y); // Error, y has no initialization inferred type
+
+
+  var z;
+  do {
+    // z is potentially assigned.
+    if (b) {
+      z = 42;
+    } else {
+      z = 37;
+    }
+  } while (something);
+  print(z); // Error, z has no initialization inferred type
+```
+
+Note that as a consequence of the above it is impossible, in an error-free
+program, to observe the static type or contents of an untyped variable before it
+has been initialized and given an inferred type.  From this standpoint it may be
```

Comment:
... unless it's `late` and only potentially assigned. Or is that covered by "error-fee"?

Reply:
I'm using initialized loosely here. I'll make it more precise.